apis/rust/node/src/config.rs

```diff
@@ -84,23 +86,62 @@ impl std::ops::Deref for DataId {
     }
 }
 
-#[derive(Debug, Clone, PartialEq, Eq)]
-pub struct InputMapping {
-    pub source: NodeId,
-    pub operator: Option<OperatorId>,
-    pub output: DataId,
+#[derive(Debug, Clone, PartialEq, Eq, PartialOrd, Ord)]
+pub enum InputMapping {
+    Timer { interval: Duration },
+    User(UserInputMapping),
+}
+
+impl InputMapping {
+    pub fn source(&self) -> &NodeId {
+        static DORA_NODE_ID: OnceCell<NodeId> = OnceCell::new();
+
+        match self {
+            InputMapping::User(mapping) => &mapping.source,
+            InputMapping::Timer { .. } => DORA_NODE_ID.get_or_init(|| NodeId("dora".to_string())),
+        }
+    }
+
+    pub fn operator(&self) -> &Option<OperatorId> {
+        match self {
+            InputMapping::User(mapping) => &mapping.operator,
+            InputMapping::Timer { .. } => &None,
+        }
+    }
+}
+
+impl ToString for InputMapping {
```

Comment:
I have read that implementing `std::fmt::Display` is the best practice and provide a larger scope than `ToString` but as we use it for serializing, maybe `ToString` is best. I would have probably implemented `Display` but you can leave it as is.

Reply:
Oh yes, good catch! Implementing `Display` is definitely better, and it should also work nicely with serde.

Reply:
Fixed in void intermediate allocations by implementing Display instead of ToString

Reply:
Fixed in 38199b4a44414b3c635de572acf66627450f4fbb